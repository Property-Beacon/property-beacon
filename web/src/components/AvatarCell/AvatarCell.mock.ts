import type { User } from 'web/types/graphql'

// Define your own mock data here:
export const standard = (input?: Partial<User>) => {
  const dateStr = new Date().toLocaleString()

  return {
    user: (input
      ? input
      : {
          id: 'a3f08caa-e057-11eb-ba80-0242ac130004',
          role: 'ADMIN',
          email: 'example@domain.com',
          logOn: dateStr,
          logOff: dateStr,
          createdAt: dateStr,
          profile: {
            id: '973cf2d6-e058-11eb-ba80-0242ac130004',
            userId: 'a3f08caa-e057-11eb-ba80-0242ac130004',
            companyId: null,
            avatar: 'https://cdn.filestackcontent.com/fpWIVZz7RSSdjY4pn5YP',
            firstName: 'David',
            lastName: 'Jones',
            phone: '0211112222',
            mobile: '0412345678',
            updatedAt: dateStr
          }
        }) as User
  }
}
