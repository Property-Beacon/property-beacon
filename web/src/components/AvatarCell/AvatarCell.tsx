import type { CellFailureProps, CellSuccessProps } from '@redwoodjs/web'
import { AiOutlineStop } from 'react-icons/ai'
import type { GetUserById } from 'types/graphql'

export const QUERY = gql`
  query GetUserById($id: String!) {
    user: getUserById(id: $id) {
      id
      role
      email
      logOn
      logOff
      createdAt
      profile {
        id
        avatar
        fullName
        firstName
        lastName
        mobile
        phone
        companyId
        updatedAt
        address {
          lat
          lng
          gPlaceId
          premise
          state
          street
          suburb
          country
          postalCode
          updatedAt
          formattedAddress
        }
      }
    }
  }
`

type Props = {
  loading?: boolean
  user?: GetUserById['user']
  error?: CellFailureProps['error']
}

const Avatar = ({ user, loading, error }: Props) => {
  const avatar = user?.profile.avatar
  const initials = user
    ? user.profile.firstName
      ? [
          (user.profile.firstName || '').charAt(0).toUpperCase(),
          (user.profile.lastName || '').charAt(0).toUpperCase()
        ].join('')
      : user.email.charAt(0).toUpperCase()
    : '?'

  return (
    <div data-testid="avatar-cell" className="avatar placeholder h-full w-full">
      {loading ? (
        <div
          data-testid="avatar-cell-loading"
          className="bg-primary-focus text-primary-content h-full w-full rounded-full animate-pulse"
        ></div>
      ) : error ? (
        <div
          data-testid="avatar-cell-error"
          className="bg-base-100 h-full w-full rounded-full"
        >
          <AiOutlineStop size={16} className="text-error" />
        </div>
      ) : avatar ? (
        <img
          alt="avatar"
          src={avatar}
          loading="lazy"
          className="rounded-full"
          data-testid="avatar-cell-img"
        />
      ) : (
        <div
          data-testid="avatar-cell-initials"
          className="bg-primary-focus text-primary-content h-full w-full rounded-full"
        >
          {initials}
        </div>
      )}
    </div>
  )
}

export const beforeQuery = (props) => {
  return {
    variables: props,
    fetchPolicy: 'cache-first',
    nextFetchPolicy: 'cache-first'
  }
}

export const Loading = () => <Avatar loading />

export const Empty = () => <Avatar />

export const Failure = ({ error }: CellFailureProps) => <Avatar error={error} />

export const Success = ({ user }: CellSuccessProps<GetUserById>) => (
  <Avatar user={user} />
)
