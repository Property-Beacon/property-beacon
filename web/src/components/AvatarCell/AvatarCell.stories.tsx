import { Empty, Failure, Loading, Success } from './AvatarCell'
import { standard } from './AvatarCell.mock'

export const loading = () => {
  return Loading ? (
    <div className="flex place-items-center">
      <div className="h-10 w-10 mr-4">
        <Loading />
      </div>
      <div className="h-12 w-12 mr-4">
        <Loading />
      </div>
      <div className="h-14 w-14 mr-4">
        <Loading />
      </div>
      <div className="h-16 w-16">
        <Loading />
      </div>
    </div>
  ) : null
}

export const empty = () => {
  return Empty ? (
    <div className="flex place-items-center">
      <div className="h-10 w-10 mr-4">
        <Empty />
      </div>
      <div className="h-12 w-12 mr-4">
        <Empty />
      </div>
      <div className="h-14 w-14 mr-4">
        <Empty />
      </div>
      <div className="h-16 w-16">
        <Empty />
      </div>
    </div>
  ) : null
}

export const failure = () => {
  return Failure ? (
    <div className="flex place-items-center">
      <div className="h-10 w-10 mr-4">
        <Failure error={new Error('oh no')} />
      </div>
      <div className="h-12 w-12 mr-4">
        <Failure error={new Error('oh no')} />
      </div>
      <div className="h-14 w-14 mr-4">
        <Failure error={new Error('oh no')} />
      </div>
      <div className="h-16 w-16">
        <Failure error={new Error('oh no')} />
      </div>
    </div>
  ) : null
}

export const success = () => {
  return Success ? (
    <div className="flex place-items-center">
      <div className="h-10 w-10 mr-4">
        <Success
          {...standard({
            email: 'example@domain.com',
            profile: {
              id: '973cf2d6-e058-11eb-ba80-0242ac130004'
            }
          })}
        />
      </div>
      <div className="h-12 w-12 mr-4">
        <Success
          {...standard({
            email: 'example@domain.com',
            profile: {
              id: '973cf2d6-e058-11eb-ba80-0242ac130004',
              firstName: 'David',
              lastName: 'Jones'
            }
          })}
        />
      </div>
      <div className="h-14 w-14 mr-4">
        <Success
          {...standard({
            email: 'example@domain.com',
            profile: {
              id: '973cf2d6-e058-11eb-ba80-0242ac130004',
              avatar: 'https://cdn.filestackcontent.com/fpWIVZz7RSSdjY4pn5YP'
            }
          })}
        />
      </div>
      <div className="h-16 w-16">
        <Success
          {...standard({
            email: 'example@domain.com',
            profile: {
              id: '973cf2d6-e058-11eb-ba80-0242ac130004',
              avatar: 'https://cdn.filestackcontent.com/fpWIVZz7RSSdjY4pn5YP'
            }
          })}
        />
      </div>
    </div>
  ) : null
}

export default { title: 'Cells/AvatarCell' }
