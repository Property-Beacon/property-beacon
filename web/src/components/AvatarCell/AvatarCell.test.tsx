import { render, screen } from '@redwoodjs/testing'
import { Empty, Failure, Loading, Success } from './AvatarCell'
import { standard } from './AvatarCell.mock'

describe('AvatarCell', () => {
  it('renders Loading successfully', () => {
    expect(() => {
      render(<Loading />)
    }).not.toThrow()
    expect(screen.getByTestId('avatar-cell-loading')).toMatchSnapshot()
  })

  it('renders Empty successfully', async () => {
    expect(() => {
      render(<Empty />)
    }).not.toThrow()
    expect(screen.getByTestId('avatar-cell-initials')).toMatchSnapshot()
  })

  it('renders Failure successfully', async () => {
    expect(() => {
      render(<Failure error={new Error('Oh no')} />)
    }).not.toThrow()
    expect(screen.getByTestId('avatar-cell-error')).toMatchSnapshot()
  })

  it('renders Success successfully', async () => {
    expect(() => {
      render(<Success {...standard()} />)
    }).not.toThrow()
    expect(screen.getByTestId('avatar-cell')).toMatchSnapshot()
  })

  it('renders initials with the first character of email address', async () => {
    render(
      <Success
        {...standard({
          email: 'david@domain.com',
          profile: {
            id: '973cf2d6-e058-11eb-ba80-0242ac130004'
          }
        })}
      />
    )
    const text = screen.getByTestId('avatar-cell-initials').textContent

    expect(text).toEqual('D')
  })

  it('renders initials with the firstName and lastName', async () => {
    render(
      <Success
        {...standard({
          email: 'david@domain.com',
          profile: {
            firstName: 'John',
            lastName: 'Smith',
            id: '973cf2d6-e058-11eb-ba80-0242ac130004'
          }
        })}
      />
    )
    const text = screen.getByTestId('avatar-cell-initials').textContent

    expect(text).toEqual('JS')
  })

  it('renders avatar image', async () => {
    render(
      <Success
        {...standard({
          email: 'david@domain.com',
          profile: {
            firstName: 'John',
            lastName: 'Smith',
            id: '973cf2d6-e058-11eb-ba80-0242ac130004',
            avatar: 'https://cdn.filestackcontent.com/fpWIVZz7RSSdjY4pn5YP'
          }
        })}
      />
    )

    expect(screen.getByTestId('avatar-cell-img')).toMatchSnapshot()
  })
})
