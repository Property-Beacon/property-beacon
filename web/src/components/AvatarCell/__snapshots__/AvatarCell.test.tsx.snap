// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AvatarCell renders Empty successfully 1`] = `
<div
  class="bg-primary-focus text-primary-content h-full w-full rounded-full"
  data-testid="avatar-cell-initials"
>
  ?
</div>
`;

exports[`AvatarCell renders Failure successfully 1`] = `
<div
  class="bg-base-100 h-full w-full rounded-full"
  data-testid="avatar-cell-error"
>
  <svg
    class="text-error"
    fill="currentColor"
    height="16"
    stroke="currentColor"
    stroke-width="0"
    viewBox="0 0 1024 1024"
    width="16"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm0 820c-205.4 0-372-166.6-372-372 0-89 31.3-170.8 83.5-234.8l523.3 523.3C682.8 852.7 601 884 512 884zm288.5-137.2L277.2 223.5C341.2 171.3 423 140 512 140c205.4 0 372 166.6 372 372 0 89-31.3 170.8-83.5 234.8z"
    />
  </svg>
</div>
`;

exports[`AvatarCell renders Loading successfully 1`] = `
<div
  class="bg-primary-focus text-primary-content h-full w-full rounded-full animate-pulse"
  data-testid="avatar-cell-loading"
/>
`;

exports[`AvatarCell renders Success successfully 1`] = `
<div
  class="avatar placeholder h-full w-full"
  data-testid="avatar-cell"
>
  <img
    alt="avatar"
    class="rounded-full"
    data-testid="avatar-cell-img"
    loading="lazy"
    src="https://cdn.filestackcontent.com/fpWIVZz7RSSdjY4pn5YP"
  />
</div>
`;

exports[`AvatarCell renders avatar image 1`] = `
<img
  alt="avatar"
  class="rounded-full"
  data-testid="avatar-cell-img"
  loading="lazy"
  src="https://cdn.filestackcontent.com/fpWIVZz7RSSdjY4pn5YP"
/>
`;
